src/portfft/descriptor.hpp

```diff
@@ -491,18 +512,89 @@ class committed_descriptor {
+  }
+
+  /**
+   * utility function to set specialization constants
+   * @tparam SubgroupSize Subgroup size
+   * @param top_level selected level of implementation
+   * @param prepared_vec vector as returned from prepare_implementation
+   * @param compute_direction direction of compute, as in forward or backward
+   * @param kernel_num which dimension are the kernels being built for
+   * @param is_compatible flag to be set if the kernels are compatible
+   * @return
+   */
+  template <Idx SubgroupSize>
+  std::vector<kernel_data_struct> set_specialization_constants(
+      detail::level top_level,
+      std::vector<std::tuple<detail::level, std::vector<sycl::kernel_id>, std::vector<Idx>>>& prepared_vec,
+      direction compute_direction, std::size_t kernel_num, bool& is_compatible) {
+    Scalar scale_factor = compute_direction == direction::FORWARD ? params.forward_scale : params.backward_scale;
+    std::size_t counter = 0;
+    bool take_conjugate_on_load = false;
+    bool take_conjugate_on_store = false;
+    std::vector<kernel_data_struct> result;
+    for (auto& [level, ids, factors] : prepared_vec) {
+      auto in_bundle = sycl::get_kernel_bundle<sycl::bundle_state::input>(queue.get_context(), ids);
+      if (top_level == detail::level::GLOBAL) {
+        if (counter == prepared_vec.size() - 1) {
+          if (compute_direction == direction::BACKWARD) {
+            take_conjugate_on_store = true;
+          }
+          set_spec_constants(
+              detail::level::GLOBAL, in_bundle,
+              static_cast<std::size_t>(std::accumulate(factors.begin(), factors.end(), Idx(1), std::multiplies<Idx>())),
+              factors, detail::elementwise_multiply::NOT_APPLIED, detail::elementwise_multiply::NOT_APPLIED,
+              detail::apply_scale_factor::APPLIED, level, take_conjugate_on_load, take_conjugate_on_store, scale_factor,
+              static_cast<Idx>(counter), static_cast<Idx>(prepared_vec.size()));
+          // reset take_conjugate_on_store
+          take_conjugate_on_store = false;
+        } else {
+          if (counter == 0 && compute_direction == direction::BACKWARD) {
+            take_conjugate_on_load = true;
+          }
+          set_spec_constants(
+              detail::level::GLOBAL, in_bundle,
+              static_cast<std::size_t>(std::accumulate(factors.begin(), factors.end(), Idx(1), std::multiplies<Idx>())),
+              factors, detail::elementwise_multiply::NOT_APPLIED, detail::elementwise_multiply::APPLIED,
+              detail::apply_scale_factor::NOT_APPLIED, level, take_conjugate_on_load, take_conjugate_on_store,
+              scale_factor, static_cast<Idx>(counter), static_cast<Idx>(prepared_vec.size()));
+          // reset take_conjugate_on_load
+          take_conjugate_on_load = false;
+        }
+      } else {
+        if (compute_direction == direction::BACKWARD) {
+          take_conjugate_on_load = true;
+          take_conjugate_on_store = true;
+        }
```

Comment:
Do these need to be reset if a non-global level is followed by the global level?

Reply:
No,
because they are set here at the beginning of the function [here](https://github.com/codeplaysoftware/portFFT/blob/a152e0de5dc11241653522368b73d62c79cc78c6/src/portfft/descriptor.hpp#L546).
Each dimension separately calls `build_w_spec_const` which would in turn call `set_specialization_constants` separately, thus there is no effect of  one implementation on the other.

Reply:
Resolving this